podman unshare: add --rootless-cni to join the ns

Luap99 opened the pull request:
Add a new --rootless-cni option to podman unshare to also join the
rootless-cni network namespace. This is useful if you want to connect
to a rootless container via IP address. This is only possible from the
rootless-cni namespace and not from the host namespace. This option also
helps to debug problems in the rootless-cni namespace.

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/master/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]".  That will prevent functional tests from running and save time and energy.
-->

mheon commented on cmd/podman/system/unshare.go:
```diff
@@ -33,6 +34,7 @@ func init() {
 	})
 	flags := unshareCommand.Flags()
 	flags.SetInterspersed(false)
+	flags.BoolVar(&unshareOptions.RootlessCNI, "rootless-cni", false, "Join the rootless-cni namespace")
```
Suggest rewording description to "Join the rootless network namespace used for CNI networking"

Luap99 replied:
Done

mheon commented on pkg/domain/infra/abi/system.go:
```diff
@@ -390,13 +390,25 @@ func unshareEnv(graphroot, runroot string) []string {
+		if err != nil {
+			return err
+		}
+		defer rootlesscni.Cleanup(ic.Libpod)
```
Is there a risk that we wipe out the rootless CNI namespace if we have two unshare sessions both joined to it, and one exits before the other?

Luap99 replied:
Yes that is problematic. What would be the proper way to fix this? Cleanup currently checks that there are no running containers with bridge networking and that `var/lib/cni/results` is empty. Otherwise cleanup is skipped.

mheon replied:
Hm. We could try reference counting, but I'm concerned we could get into a state where we never clean up if a `podman unshare` gets SIGKILL'd before it can indicate it's done.

The easiest way is probably to guarantee there's only 1 `podman unshare` shell with `--rootless-cni` active at a time using a lockfile. I don't think we'll run into cases where more than 1 session is needed very often?

Luap99 replied:
Yeah I think we should ensure that we always do cleanup. Otherwise we will not only leak the namespace but also the slirp process.

Only allowing one session feels a little odd. The only thing that will happen to the other unshare session is that we lose network connectivity because we kill the slirp process. The namespace stays alive until the process exits only the bind mount is removed bind cleanup.  Also as long as at least one container with CNI networking is running we can safely use as many sessions as we want.

I think we can live with this problem.

mheon replied:
OK, that ought to be fine, then.

giuseppe replied:
could we just join the network namespace without creating it?

That would also solve the issue that we don't have to cleanup on exit?

Luap99 replied:
We could do this but it doesn't solve the problem. We could skip the cleanup here but if all running cni containers are stopped we will perform the cleanup and then all running unshare sessions would lose network connectivity.
Also the reason I choose to create the ns here is because this should open up more use cases. I thought about using this to create the ns and do some setup stuff, e.g. create a interface used by cni.

giuseppe replied:
could we split the two cases, perhaps `--rootless-cni=create|join`?  I think having an option to just join the network namespace, and lose connectivity if there are no other containers running is still a valid use case.

Luap99 replied:
There is only one rootless cni namespace per user. We will always join if it exists.

mheon commented:
Compose test failing, potential flake? Restarted.

LGTM

Luap99 commented:
> Compose test failing, potential flake? Restarted.

Yes I have to fix this ASAP

mheon commented:
Needs a rebase

Luap99 commented:
Updated. I added a note to the man page that using more than one session can have unexpected results. I also added a extra check to prevent a possible panic.

TomSweeneyRedHat commented:
LGTM
and happy green test buttons.

rhatdan commented:
/lgtm
